caravel/assets/javascripts/explorev2/reducers/exploreReducer.js

```diff
@@ -17,55 +26,163 @@ export const exploreReducer = function (state, action) {
     },
     [actions.SET_GROUPBY_COLUMNS]() {
-      return Object.assign({}, state, { groupByColumns: action.groupByColumns });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            groupByColumns: action.groupByColumns,
+          },
+        },
+      };
     },
     [actions.SET_METRICS_OPTS]() {
       return Object.assign({}, state, { metricsOpts: action.metricsOpts });
     },
     [actions.SET_METRICS]() {
-      return Object.assign({}, state, { metrics: action.metrics });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            metrics: action.metrics,
+          },
+        },
+      };
     },
     [actions.SET_COLUMN_OPTS]() {
       return Object.assign({}, state, { columnOpts: action.columnOpts });
     },
     [actions.SET_NOT_GROUPBY_COLUMNS]() {
-      return Object.assign({}, state, { columns: action.columns });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            columns: action.columns,
+          },
+        },
+      };
     },
     [actions.SET_ORDERING_OPTS]() {
       return Object.assign({}, state, { orderingOpts: action.orderingOpts });
     },
     [actions.SET_ORDERINGS]() {
-      return Object.assign({}, state, { orderings: action.orderings });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            orderings: action.orderings,
+          },
+        },
+      };
     },
     [actions.SET_TIME_STAMP_FORMAT]() {
-      return Object.assign({}, state, { timeStampFormat: action.timeStampFormat });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            timeStampFormat: action.timeStampFormat,
+          },
+        },
+      };
     },
     [actions.SET_ROW_LIMIT]() {
-      return Object.assign({}, state, { rowLimit: action.rowLimit });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            rowLimit: action.rowLimit,
+          },
+        },
+      };
     },
     [actions.TOGGLE_SEARCHBOX]() {
       return Object.assign({}, state, { searchBox: action.searchBox });
     },
     [actions.SET_WHERE_CLAUSE]() {
-      return Object.assign({}, state, { whereClause: action.whereClause });
+      return {
+        ...state,
+        viz: {
+          ...state.viz,
+          formData: {
+            ...state.viz.formData,
+            where: action.where,
+          },
+        },
+      };
     },
     [actions.SET_HAVING_CLAUSE]() {
-      return Object.assign({}, state, { havingClause: action.havingClause });
+      return {
```

Comment:
looks like a lot of code repitition, what about defining the function setFormData that will modify the state for the form data
`return setFormData({having: action.having})`